### PR TITLE
New: Add consistent and ..-as-needed to object-shorthand (fixes #5438)

### LINKED ISSUE #5438: Proposal addition to object-shorthand: add consistent and consistent-as-needed
**When does this rule warn? Please describe and show example code:**
Similar to the `quote-props` rule, where it's possible to set the rule to `consistent` and `consistent-as-needed`, I'd like the `object-shorthand` rule to include these options as well.

When `object-shorthand` set to `consistent`, the below should not be considered a problem:
```js
let obj1 = {
    foo: foo,
    bar: 'This is a string',
};
```
No problem:
```js
let obj2 = {
    foo,
    bar,
};
```

The following should be considered a problem:
```js
let obj1 = {
    foo,
    bar: 'This is a string',
};
```

Problem:
```js
let obj2 = {
    a() {},
    bar: foo,
};
```

The `consistent-as-needed` is very similar, but if all keys are redundant, the following pattern is considered a problem:
```js
let obj3 = {
    foo: foo,
    bar: bar,
};
```

The options apply to both properties and methods.

**Is this rule preventing an error or is it stylistic?**
Stylistic. Personally, I feel the shorthand syntax is great, but mixing it with the longform syntax makes it harder to scan/read code. In my rules, I want to allow the shorthand syntax, but only if it's consistent in that object literal.

**Why is this rule a candidate for inclusion instead of creating a custom rule?**
I think this is in line with other rules which also provide consistent and consistent-as-needed options (e.g. `quote-props`).

**Are you willing to create the rule yourself?**
Yes.

### PR DESCRIPTION
**Pull request by martijndeh**
Please review my implementation for https://github.com/eslint/eslint/issues/5438.

### CODE REVIEW

**Comment by alberto** [comment]
Marked as `do not merge` since the issue is not accepted yet.

**Comment by nzakas** [comment]
@martijndeh if you'd like to rebase, we can start moving this forward.

**Comment by IanVS** [comment]
@martijndeh If you're willing to rebase this, I'd love for us to get this merged.

**Comment by martijndeh** [comment]
@IanVS Yes, will do. Thank you.

**Comment by kaicataldo** [comment]
@martijndeh Pinging one more time since it looks like you did some great work on this!

**Comment by martijndeh** [comment]
I just rebased. Let me know if this needs changing. This is the first time I touched the eslint codebase. :)

File: docs/rules/object-shorthand.md [hunk 1]
```diff
@@ -152,6 +154,54 @@ var foo = {
...
+/*eslint object-shorthand: [2, "consistent"]*/
+/*eslint-env es6*/
+
+var foo = {
+    a: a,
+    b: 'foo'
+};
+```
+
+The following will also *not* warn:
```
Comment (nzakas) [inline docs/rules/object-shorthand.md]: You can combine this with the preceding example. No need to have two code blocks.

File: docs/rules/object-shorthand.md [hunk 2]
```diff
@@ -152,6 +154,54 @@ var foo = {
...
+
+The following will *not* warn:
+
+```js
+/*eslint object-shorthand: [2, "consistent"]*/
+/*eslint-env es6*/
+
+var foo = {
+    a: a,
+    b: 'foo'
```
Comment (nzakas) [inline docs/rules/object-shorthand.md]: Nitpick: we use double quotes in examples.

File: lib/rules/object-shorthand.js [hunk 3]
```diff
@@ -117,11 +131,86 @@ module.exports = {
...
+                    if (shorthandProperties.length > 0) {
+                        context.report(node, "Inconsistently using longform and shortform syntax.");
+                    } else if (checkRedundancy) {
+
+                        // If all properties of the object contain a method or value with a name matching it's key,
+                        // all the keys are redudant.
+                        canAlwaysUseShorthand = properties.every(isRedudant);
+
+                        if (canAlwaysUseShorthand) {
+                            context.report(node, "Properties shouldn't be in longform as all keys are redundant.");
```
Comment (nzakas) [inline lib/rules/object-shorthand.js]: We usually put messages in the format "Expected x" or "Expected but found y.", So maybe "Expected shorthand for all properties."?

File: lib/rules/object-shorthand.js [hunk 4]
```diff
@@ -117,11 +131,86 @@ module.exports = {
...
+                shorthandProperties = properties.filter(isShorthand);
+
+                // If we do not have an equal number of longform properties as
+                // shorthand properties, we are using the annotations inconsistently
+                if (shorthandProperties.length !== properties.length) {
+
+                    // We have at least 1 shorthand property
+                    if (shorthandProperties.length > 0) {
+                        context.report(node, "Inconsistently using longform and shortform syntax.");
+                    } else if (checkRedundancy) {
```
Comment (nzakas) [inline lib/rules/object-shorthand.js]: Maybe "Unexpected mix of shorthand and non-shorthand properties."?

**Comment by nzakas** [comment]
Overall looks good, just a bit of refinement in the messaging.

**Comment by martijndeh** [comment]
Thanks for the comments. I've replaced the quotes and refined the error messages.

**Comment by nzakas** [comment]
@martijndeh it looks like there are several tests failing now. Can you take a look?

**Comment by nzakas** [comment]
@martijndeh following up, do you intend to finish this up? There are still several tests failing

**Comment by martijndeh** [comment]
@nzakas Yes, I will pick this up later this week.

**Comment by martijndeh** [comment]
@nzakas Everything should be resolved now. Let me know if there is something else I can do.

**Comment by martijndeh** [comment]
@nzakas I resolved some new conflicts again. I see you're moving to ES6 and noticed the var-to-let refactor. I took the liberty to refactor the object-shorthand rule to use const where possible, even though `prefer-const` is not enabled (yet). I hope this is ok.

**Comment by nzakas** [comment]
Yup that's fine. Can I trouble you to rebase one more time? We've had a lot of changes lately. (Last time, I promise!)

**Comment by martijndeh** [comment]
Sure no problem. I'll have time later this week.

**Comment by martijndeh** [comment]
@nzakas I just did a rebase again.

**Comment by ilyavolodin** [comment]
LGTM. Thanks!

**Pull request merged**